### PR TITLE
Download pricing data in a goroutine to prevent deadlock when updating Azure configuration

### PR DESCRIPTION
**Pull request by nikovacevic**
## What does this PR change?
* Fixes an issue where updating configuration (e.g. currency) on Azure would deadlock. Now, it doesn't deadlock, and the configuration is saved.

## How will this PR impact users?
* It will let them update their configuration!

## Does this PR address any GitHub or Zendesk issues?
* Closes https://github.com/kubecost/kubecost-cost-model/issues/992

## How was this PR tested?
* Manually, on `khand-dev-1` where repro steps were well documented and easy to follow! (Thank you @ivankube.) Before, the request never completed and the handler got deadlocked. After, everything works as intended.

### CODE REVIEW

File: pkg/cloud/azureprovider.go [hunk 1]
```diff
@@ -1208,35 +1210,46 @@ func (az *Azure) UpdateConfig(r io.Reader, updateType string) (*CustomPricing, e
...
 			c.AzureCloud = asc.AzureCloud
 		} else {
-			defer az.DownloadPricingData()
+			// This will block if not in a goroutine. It calls GetConfig(), which
+			// in turn calls GetCustomPricingData, which acquires the same lock
+			// that is acquired by az.Config.Update, which is the function to
+			// which this function gets passed, and subsequently called. Booo.
+			defer func() {
+				go az.DownloadPricingData()
+			}()
```
Comment (nikovacevic) [inline pkg/cloud/azureprovider.go:1221]: This is the fix. Everything else is error wrapping, gofmt, etc.

**Review by Sean-Holcomb — APPROVED** [review]
looks good!

**Review by AjayTripathy — APPROVED** [review]

**Comment by AjayTripathy** [comment]
Can we make sure this gets cherrypicked to the release branch as well?

File: pkg/cloud/azureprovider.go (hunk 1 above)
Comment (mbolt35) [reply pkg/cloud/azureprovider.go:1221]: Not sure how to best protect ourselves against reentrant locking, or why golang doesn't support this. Provider APIs need a major face lift.

File: pkg/cloud/azureprovider.go [hunk 2]
```diff
@@ -19,6 +19,8 @@ import (
 	"github.com/opencost/opencost/pkg/util"
 	"github.com/opencost/opencost/pkg/util/fileutil"
 	"github.com/opencost/opencost/pkg/util/json"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
```
Comment (mbolt35) [inline pkg/cloud/azureprovider.go:23]: What's up with these imports?

**Review by mbolt35 — CHANGES_REQUESTED** [review]
Looks good to me outside of the imports. Maybe accidental import while debugging?

File: pkg/cloud/azureprovider.go (hunk 2 above)
Comment (nikovacevic) [reply pkg/cloud/azureprovider.go:23]: Go is telling me that `strings.Title` is deprecated. This is the alternative. See:

```go
- kUpper := strings.Title(k)
```
```go
+ kUpper := cases.Title(language.Und, cases.NoLower).String(k)
```

Comment (nikovacevic) [reply pkg/cloud/azureprovider.go:23]: Honestly... could've just left it. But I can't resist a little clean up effort when it's there.

**Comment by nikovacevic** [comment]
Cherry-picked here: https://github.com/opencost/opencost/pull/1450

File: pkg/cloud/azureprovider.go (hunk 2 above)
Comment (mbolt35) [reply pkg/cloud/azureprovider.go:23]: Can we, at least do:
```
var toTitle cases.Caser = cases.Title(language.Und, cases.NoLower)

// ...

kUpper := toTitle.String(k)
```

I think this is the intentional way to use it? The title caser creation doesn't look like we should have to create it every time we want a title.

**Comment by michaelmdresser** [comment]
Just a heads up here -- the cherry pick PR of this change is already merged. Let's be careful to not let things get out of sync if this PR changes.

**Comment by nikovacevic** [comment]
I just need Bolt to hit the big green button

File: pkg/cloud/azureprovider.go (hunk 2 above)
Comment (nikovacevic) [reply pkg/cloud/azureprovider.go:23]: Yep! Done, pushing now.

**Review by mbolt35 — APPROVED** [review]

**Pull request merged**